### DISCUSSION TITLE
User-agent matchmaking

### DISCUSSION THREAD

**Discussion by pchampin** [Wimmics/solid-indexer#2] [General]
How does a user *start* the indexer agent?

Here are 3 possible scenarios:

1. The agent program is manually started by the user, with some configuration indicating the URL of the user's profile (the agent is then able to discover the storage URL, and the jobs assigned to it by the user, as [I proposed](https://github.com/Wimmics/solid-indexer/discussions/1#discussioncomment-5659029) in #1). This solution is only acceptable for tech-savvy users, so we should eventually explore the other options below -- but in the time frame of the hackathon, we will probably not have the time to go further?

2. The agent program runs permanently on its own server, and has its own WebID and its own POD (or at least an Solid inbox). In order to "invite" the agent to execute the jobs it has created, the user-facing application will post an invitation in the agent's inbox, indicating the URL of the user's profile. This solution could be deployed with the existing Solid spec, but it requires that the user trusts the agent's server -- e.g. that the server does not run a malicious version of the agent that leaks the user's data to third parties.

3. In addition to the identity server and the storage server, Solid could standardize another kind of server ("execution server") that would allow users to run code on their behalf in the background (similar to JS workers in Web apps, but server-side). The agent would then be executed on this server, which the user is assumed to trust (as much as they trust their storage server). How the user starts and stops programs on this "execution server" would need to be specified in the Solid protocol.

**Comment by lecoqlibre** [comment]
2. This solution would require a server-to-server mechanism to get the agent notified when a new invite comes into its inbox. We can't use web-sockets (browser-to-server only). Instead, we could use web-hooks to notify the agent it has jobs to do on the user's POD. Note that to use web-hooks we need to define endpoint(s) that will be called by the client. This would bring us to the client-to-client standard field. We might use ActivityPub to get a more standard fashion to use web-hooks.
3. That makes me think about linux distribution packaging systems and repositories. A server could provide a store of trusted applications that a user can "install" on its POD. This need of trust should be a good point for open-source apps.
